internal/api/permissions.go

```diff
@@ -72,6 +90,189 @@ func (r *Router) checkAction(c echo.Context) error {
+		return echo.NewHTTPError(http.StatusBadRequest, "failed to get the subject").SetInternal(err)
+	}
+
+	subjectResource, err := r.engine.NewResourceFromID(subject)
+	if err != nil {
+		return echo.NewHTTPError(http.StatusBadRequest, "error processing subject ID").SetInternal(err)
+	}
+
+	var reqBody checkPermissionsRequest
+
+	if err := c.Bind(&reqBody); err != nil {
+		return echo.NewHTTPError(http.StatusBadRequest, "error parsing request body").SetInternal(err)
+	}
+
+	var errs []error
+
+	requestsCh := make(chan checkRequest, len(reqBody.Actions))
+
+	for i, check := range reqBody.Actions {
+		if check.Action == "" {
+			errs = append(errs, fmt.Errorf("check %d: %w", i, ErrNoActionDefined))
+
+			continue
+		}
+
+		resourceID, err := gidx.Parse(check.ResourceID)
+		if err != nil {
+			errs = append(errs, fmt.Errorf("check %d: %w: error parsing resource id: %s", i, err, check.ResourceID))
+
+			continue
+		}
+
+		resource, err := r.engine.NewResourceFromID(resourceID)
+		if err != nil {
+			errs = append(errs, fmt.Errorf("check %d: %w: error creating resource from id: %s", i, err, resourceID.String()))
+
+			continue
+		}
+
+		requestsCh <- checkRequest{
+			Index:    i,
+			Resource: resource,
+			Action:   check.Action,
+		}
+	}
+
+	close(requestsCh)
+
+	if len(errs) != 0 {
+		return echo.NewHTTPError(http.StatusBadRequest, "invalid check request").SetInternal(multierr.Combine(errs...))
+	}
+
+	doneCh := make(chan bool)
+	resultsCh := make(chan checkResult, len(reqBody.Actions))
+
+	ctx, cancel := context.WithTimeout(ctx, maxCheckDuration)
+
+	defer cancel()
+
+	for i := 0; i < r.concurrentChecks; i++ {
+		go func() {
+			for check := range requestsCh {
+				result := &checkResult{
+					Request: check,
+				}
+
+				// Check the permissions
+				err := r.engine.SubjectHasPermission(ctx, subjectResource, check.Action, check.Resource)
+				if err != nil {
+					result.Error = err
+				}
+
+				// Check if doneCh has been closed, if so, don't write to resultsCh.
+				select {
+				case <-doneCh:
+					return
+				default:
+				}
+
+				resultsCh <- *result
+			}
+		}()
+	}
+
+	var (
+		unauthorizedErrors int
+		internalErrors     int
+		allErrors          []error
+	)
+
+	go func() {
+		var count int
+
+		for result := range resultsCh {
+			count++
+
+			if result.Error != nil {
+				if errors.Is(result.Error, query.ErrActionNotAssigned) {
+					err := fmt.Errorf("%w: subject '%s' does not have permission to perform action '%s' on resource '%s'",
+						ErrAccessDenied, subject, result.Request.Action, result.Request.Resource.ID.String())
+
+					unauthorizedErrors++
+
+					allErrors = append(allErrors, err)
+				} else {
+					err := fmt.Errorf("check %d: %w", result.Request.Index, result.Error)
+
+					internalErrors++
+
+					allErrors = append(allErrors, err)
+				}
+
+				close(doneCh)
+				close(resultsCh)
+
+				return
+			}
+
+			if count == len(reqBody.Actions) {
+				close(doneCh)
+				close(resultsCh)
+			}
+		}
+	}()
```

Comment:
Rather than ranging over the channel, if we know how many results we expect back we should be able to just read those. Something like this:

```go
for i := 0; i < numChecks; i++ {
  select {
    case result := <-resultsCh:
      // check for errors
    case <-ctx.Done():
      // context error handling
  }
}
```

Doing this would let us take out `doneCh` completely and remove the need to process results in a separate goroutine from the rest of the handler.

Reply:
ah yes, i see what you mean. it will just populate the errors with the same context error for all the ones which were cancelled. makes sense!